src/components/swap/SwapBuyFiatButton.tsx

```diff
@@ -0,0 +1,139 @@
+import { Trans } from '@lingui/macro'
+import { useWeb3React } from '@web3-react/core'
+import { ButtonText } from 'components/Button'
+import { MouseoverTooltipContent } from 'components/Tooltip'
+import { useWalletDrawer } from 'components/WalletDropdown'
+import { useCallback, useEffect, useState } from 'react'
+import { useBuyFiatFlowCompleted } from 'state/user/hooks'
+import styled, { useTheme } from 'styled-components/macro'
+import { ExternalLink } from 'theme'
+
+import { useFiatOnrampAvailability, useOpenModal } from '../../state/application/hooks'
+import { ApplicationModal } from '../../state/application/reducer'
+
+const Dot = styled.div`
+  height: 8px;
+  width: 8px;
+  background-color: ${({ theme }) => theme.accentActive};
+  border-radius: 50%;
+`
+
+const MOONPAY_REGION_AVAILABILITY_ARTICLE =
+  'https://support.uniswap.org/hc/en-us/articles/11306664890381-Why-isn-t-MoonPay-available-in-my-region-'
+
+enum BuyFiatFlowState {
+  // Default initial state. User is not actively trying to buy fiat.
+  INACTIVE,
+  // Buy fiat flow is active and region availability has been checked.
+  ACTIVE_POST_REGION_CHECK,
+  // Buy fiat flow is active, feature is available in user's region & needs wallet connection.
+  ACTIVE_NEEDS_WALLET_CONNECTION,
+}
+
+export default function SwapBuyFiatButton() {
+  const theme = useTheme()
+  const { account } = useWeb3React()
+  const openFiatOnRampModal = useOpenModal(ApplicationModal.FIAT_ONRAMP)
+  const [buyFiatFlowCompleted, setBuyFiatFlowCompleted] = useBuyFiatFlowCompleted()
+  const [checkFiatRegionAvailability, setCheckFiatRegionAvailability] = useState(false)
+  const {
+    available: fiatOnrampAvailable,
+    availabilityChecked: fiatOnrampAvailabilityChecked,
+    error,
+    loading: fiatOnrampAvailabilityLoading,
+  } = useFiatOnrampAvailability(checkFiatRegionAvailability)
+  const [buyFiatFlowState, setBuyFiatFlowState] = useState(BuyFiatFlowState.INACTIVE)
+  const [walletDrawerOpen, toggleWalletDrawer] = useWalletDrawer()
+
+  /* 
+  Once the buy fiat flow has been completed, reset BuyFiatFlowState and also save to 
+  local storage that user has completed this flow already.
+  */
+  const terminateBuyFiatFlow = useCallback(() => {
```

Comment:
@tinaszheng okayyy so separating this out into another useEffect actually introduced more bugs (the blue dot wasn't appearing when it should), because setState doesn't happen IMMEDIATELY in react (they're batched for perf reasons). Also, having this called in another useEffect introduced the need for another "terminal" state in BuyFiatFlowState (eg: when either the modal is available and ready to be shown to the user OR when it is unavailable). 

Thus I think using another function expressively called "terminateBuyFiatFlow" is better than using another useEffect.